src/mgr/DaemonServer.cc

```diff
@@ -1899,6 +2094,107 @@ bool DaemonServer::_handle_command(
       cmdctx->reply(0, ss);
     }
     return true;
+  } else if (prefix == "osd ok-to-upgrade") {
+    std::string crush_bucket_name;
+    cmd_getval(cmdctx->cmdmap, "crush_bucket", crush_bucket_name);
+    std::string ceph_version;
+    cmd_getval(cmdctx->cmdmap, "ceph_version", ceph_version);
+    int64_t max = 1;
+    cmd_getval(cmdctx->cmdmap, "max", max);
+    int r;
+    std::set<int> osds_in_crush_bucket;
+    // Validate the crush bucket name & type. For this command the
+    // bucket type is limited to 'rack', 'chassis', 'host' or 'osd'.
+    // This is to help limit the number of OSDs and avoid
+    // performance issues during the upgrade check.
+    cluster_state.with_osdmap([&](const OSDMap& osdmap) {
+        r = osdmap.get_osds_by_bucket_name(crush_bucket_name,
+                                           &osds_in_crush_bucket);
+        // Validate crush bucket
+        if (r == -ENOENT) {
+          ss << "\"" << crush_bucket_name << "\" does not exist";
+          return;
+        } else if (r < 0) {
+          ss << "cannot parse bucket name:\"" << crush_bucket_name << "\"";
```

Comment:
nit; printing `r` might be really helpful when dealing with unforeseen problems in the wild.